File: test-samples/sample.php
<?php
// ============================================
// Theme Check: PHP
// ============================================
// - Keywords: class, function, interface, trait, namespace
// - Variables ($var), properties, visibility modifiers
// - Type declarations and nullable types
// - Comments (single line, block, docblock)
// - Operators / Numbers / Strings (single, double, heredoc)
// - Match expression / arrow functions
// - Exception handling

declare(strict_types=1);

namespace App\Models;

use App\Contracts\Identifiable;
use Exception;

// --- Constants ---
const MAX_RETRIES = 3;
const PI = 3.14159;
const HEX_VALUE = 0xFF;
const BINARY = 0b1010;
const GREETING = "Hello, World!";

// --- Interface ---
interface Area
{
    public function area(): float;
    public function name(): string;
}

// --- Trait ---
trait Timestamps
{
    protected ?string $createdAt = null;

    public function touch(): void
    {
        $this->createdAt = date('Y-m-d H:i:s');
    }
}

/**
 * Represents a user in the system.
 *
 * @property int    $id
 * @property string $name
 */
class User implements Identifiable
{
    use Timestamps;

    public function __construct(
        public readonly int $id,
        public string $name,
        private string $email,
        public bool $isActive = true,
        public array $tags = [],
    ) {
        $this->email = strtolower($email);
    }

    public function getDisplayName(): string
    {
        return "{$this->name} <{$this->email}>";
    }

    public static function validateEmail(string $email): bool
    {
        return str_contains($email, '@') && str_contains($email, '.');
    }
}

// --- Enum (PHP 8.1+) ---
enum Status: string
{
    case Active = 'active';
    case Inactive = 'inactive';
    case Pending = 'pending';

    public function label(): string
    {
        return match ($this) {
            Status::Active => 'Active user',
            Status::Inactive => 'Disabled user',
            Status::Pending => 'Awaiting approval',
        };
    }
}

// --- Generic-ish repository ---
class Repository
{
    /** @var array<int, User> */
    private array $items = [];

    public function add(int $id, User $item): void
    {
        $this->items[$id] = $item;
    }

    public function get(int $id): ?User
    {
        return $this->items[$id] ?? null;
    }

    public function all(): array
    {
        return array_values($this->items);
    }
}

// --- Arrow functions & higher-order ---
$transform = fn (int $x): int => $x * 2 + 1;
$numbers = array_map($transform, range(0, 4));
$filtered = array_filter($numbers, fn ($x) => $x > 5);

// --- String types ---
$rawString = 'no $interpolation here';
$doubleString = "Interpolated: {$numbers[0]}";
$heredoc = <<<EOT
This is a
multiline heredoc with $rawString
EOT;
$nowdoc = <<<'EOT'
No $interpolation in nowdoc
EOT;

// --- Exception handling ---
try {
    $value = (int) "not a number";
    if ($value === 0) {
        throw new Exception("Invalid value");
    }
} catch (Exception $e) {
    echo "Error: {$e->getMessage()}" . PHP_EOL;
} finally {
    echo "Cleanup" . PHP_EOL;
}

// --- Main ---
$user = new User(1, "Alice", "Alice@Example.com");
echo $user->getDisplayName() . PHP_EOL;
echo Status::Active->label() . PHP_EOL;
var_dump($user instanceof User);

File: test-samples/sample.rb
# ============================================
# Theme Check: Ruby
# ============================================
# - Keywords: def, class, module, do/end, require
# - Symbols (:symbol) and instance/class variables (@var, @@var)
# - Comments (single line, =begin/=end block)
# - Operators / Numbers / Strings (single, double, heredoc)
# - Blocks, procs, lambdas
# - String interpolation and regex
# - Exception handling

require 'set'
require 'json'

# --- Constants ---
MAX_RETRIES = 3
PI = 3.14159
HEX_VALUE = 0xFF
BINARY = 0b1010
GREETING = "Hello, World!"

# --- Module (mixin) ---
module Timestamps
  attr_reader :created_at

  def touch
    @created_at = Time.now
  end
end

=begin
Represents a user in the system.
Demonstrates attr accessors, keyword args, and mixins.
=end
class User
  include Timestamps

  attr_accessor :name
  attr_reader :id, :email

  @@count = 0

  def initialize(id:, name:, email:, is_active: true, tags: [])
    @id = id
    @name = name
    @email = email.downcase
    @is_active = is_active
    @tags = tags
    @@count += 1
  end

  def display_name
    "#{@name} <#{@email}>"
  end

  def self.count
    @@count
  end

  def self.validate_email(email)
    email.include?("@") && email.include?(".")
  end

  def active?
    @is_active
  end

  def to_s
    display_name
  end
end

# --- Class with comparable ---
class Repository
  include Enumerable

  def initialize
    @items = {}
  end

  def add(id, item)
    @items[id] = item
  end

  def get(id)
    @items.fetch(id, nil)
  end

  def each(&block)
    @items.values.each(&block)
  end
end

# --- Blocks, procs, lambdas ---
transform = ->(x) { x * 2 + 1 }
square = proc { |x| x ** 2 }
numbers = (0..4).map { |x| transform.call(x) }
filtered = numbers.select { |x| x > 5 }
total = numbers.reduce(0) { |sum, x| sum + x }

# --- Symbols & hashes ---
config = {
  host: "localhost",
  port: 8080,
  enabled: true,
}
status_labels = {
  active: "Active user",
  inactive: "Disabled user",
}

# --- String types & interpolation ---
raw_string = 'no #{interpolation} here'
double_string = "Result: #{42 * 2}"
heredoc = <<~TEXT
  This is a
  multiline heredoc with #{GREETING}
TEXT

# --- Regex ---
email_pattern = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
puts "valid" if "alice@example.com" =~ email_pattern

# --- Case / when ---
def handle_command(command)
  case command
  when "quit", "exit"
    "Goodbye"
  when "help"
    "Available commands: quit, help, status"
  when /\Aecho /
    command[5..]
  else
    "Unknown: #{command}"
  end
end

# --- Exception handling ---
begin
  value = Integer("not a number")
rescue ArgumentError => e
  puts "Error: #{e.message}"
rescue TypeError
  puts "Type error"
else
  puts "Success"
ensure
  puts "Cleanup"
end

# --- Main guard ---
if __FILE__ == $PROGRAM_NAME
  user = User.new(id: 1, name: "Alice", email: "Alice@Example.com")
  puts user.display_name
  puts user.is_a?(User)
  puts User.count
end
